File: Native Mobile Programming/Lab8/SharedPreferences.java/app/src/main/java/com/afiq/sharedpreferences/LoginViewModel.java
package com.afiq.sharedpreferences;

import android.app.Application;
import android.content.Context;
import android.content.SharedPreferences;
import android.util.Pair;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

public class LoginViewModel extends AndroidViewModel {
    private final SharedPreferences sharedPreferences;
    public LoginViewModel(@NonNull Application application) {
        super(application);
        sharedPreferences = application.getSharedPreferences("user_data", Context.MODE_PRIVATE);
    }
    public Pair<String, String> getCredentials(){
        String username = sharedPreferences.getString("username",null);
        String password = sharedPreferences.getString("password",null);
    return (username !=null && password!= null)?new Pair<>(username,password):null;
    }

    public void saveData(String username, String password){
        if((!username.isEmpty())&&(!password.isEmpty())){
            sharedPreferences.edit().putString("username", username).apply();
            sharedPreferences.edit().putString("password",password).apply();
        }else{
            System.out.println("LoginViewModel Can not save empty string");
        }
    }
    public void clearData(){
        sharedPreferences.edit().remove("username").remove("password").apply();
    }

    public static class Factory implements ViewModelProvider.Factory{
        private final Application application;

        public Factory(Application application){
            this.application = application;
        }

        @NonNull
        @Override
        public <T extends ViewModel>T create(@NonNull Class<T> modelClass){
            if(modelClass.isAssignableFrom(LoginViewModel.class)){
                return (T) new LoginViewModel(application);
            }
            throw new IllegalArgumentException();
        }
    }
}

File: Native Mobile Programming/Lab8/SharedPreferences.kt/app/src/main/java/com/afiq/sharedpreferenceskotlin/LoginViewModel.kt
package com.afiq.sharedpreferenceskotlin

import android.app.Application
import android.content.Context
import android.content.SharedPreferences
import androidx.core.content.edit
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import java.lang.IllegalArgumentException

class LoginViewModel(application: Application) : AndroidViewModel(application) {

    private val sharedPreferences: SharedPreferences by lazy {
        application.getSharedPreferences("login_data", Context.MODE_PRIVATE)
    }

    fun getCredentials(): Pair<String, String>? {
        val username = sharedPreferences.getString("username", null)
        val password = sharedPreferences.getString("password", null)
        return if (username != null && password != null) {
            Pair(username, password)
        } else {
            null
        }
    }

    fun saveSignUpData(username: String, password: String) {
        sharedPreferences.edit {
            if(username.isNotEmpty()&& username.isNotBlank()
                &&password.isNotEmpty()&& password.isNotBlank()){
                println("username =$username and password = $password")
                putString("username", username)
                putString("password", password)
            }else{
                println("LoginViewModel Data is Empty")
            }

        }
    }
    fun clearCredentials() {
        sharedPreferences.edit {
            remove("username")
            remove("password")
        }
    }


    class Factory(private val application: Application) : ViewModelProvider.Factory {
        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            if (modelClass.isAssignableFrom(LoginViewModel::class.java)) {
                @Suppress("UNCHECKED_CAST")
                return LoginViewModel(application) as T
            }
            throw IllegalArgumentException("Unknown ViewModel class")
        }
    }
}

